### PR TITLE
44 FIX - Improve spec validation messages

### PR DESCRIPTION
**Pull request by fredbi**
**The work is done. Please review and, if agreed, merge.**

This P.R is about solving open issues with the spec validation package.
[check boxes show current progress on my dev - not necessarily commited]

Proposal for breaking up:

## Scope for this P.R
- Fix the following issues
  - [x] Issue go-openapi/validate#16 - Cached compiled regexp
  - [x] Issue go-openapi/validate#12 - Multiple instances of messages
  - [x] Issue go-swagger/go-swagger#1307 - Multiple instances of messages (duplicate of go-openapi/validate#12)
  - [x] Issue go-swagger/go-swagger#1050 - Variant in path params (e.g. dot separated)
  - [x] Issue go-swagger/go-swagger#1243 - Explicit error for $ref within headers
  - [x] Issue go-swagger/go-swagger#161 - parameter items validation (a default value check)
  - [x] Issue go-swagger/go-swagger#581 - Guard int values against overflow in constraints (min/max 
        /multipleOf). Partial coverage: only together with default value check at this stage.
  - [x] Issue go-swagger/go-swagger#859 - Richer message for unresolvable $ref
        (better than "object has no key").
  
- Add the following validation features (not from issues):
  - [x] Improve checks on garbled path params ({x/y} , ..)
  - [x] Check for path param uniqueness
  - [x] Check path params are declared as required
  - [x] Optional "continue on error" mode, to allow for checks to continue and sweep through all 
          detected errors. This mode would be activated in swagger validate, but by default set to false
          for faster validation.
  - [x] Continue on error mode requires stricter guards against panics and attempts to range through nils
  - [x] Replace regexp.MustCompile based on external output by simple compile and error management
  - [x] More unit testing, with stricter assertions on errors, warnings and their numbers
  - [x] A nice to have one: improved selection of relevant messages in the OneOf|AnyOf selection, 
        and avoid annoying and unnatural "cannot validate OneOf()" while missing important points
        (as underlying messages are filtered).
  - [x] Added check on multipleOf factor which must be positive
  - [x] Easier warnings management with both errors and warnings in the same Result and methods to 
         merge errors as warnings and conversely

## Planned batch for next P.R(s)
I picked up three "hard problems", then 3 "average" but no so easy issues.
  - [ ] Issue go-swagger/go-swagger#1231 - Examples not validated.
        This feature has absolutely no support in validate, with the exception of the Examples section in responses. It should follow how we validate default values.
  - [ ] Issue go-swagger/go-swagger#581 - Guard int values against overflow in constraints (min/max/multipleOf).
        This would provide spec consistency between object/param types and constraints, and avoid possible loss of information when casting int64 to float64. Don't be mistaken, it is *not* about providing bigint arithmetics. *Current progress status*: made it work on inline params (validator.go), but will be long to generalize to all situations (objects, arrays).
  - [ ] Issue go-swagger/go-swagger#121 - Not solved with line numbers, but improve context reporting using key paths.
        The idea is to extend the Result object to provide a map of paths linked to errors.
        Many changes since this would affect the interface with Result (AddError and possibly some more direct usage).
        In the end, messages may be pretty printed by sorting and indenting keys. Not quite line numbers but better than the unordered list we have today.
  - [ ] Issue go-swagger/go-swagger#1289 - Explicit error for $ref with siblings (harder than expected!)
  - [ ] Remove explicit panics when avoidable
  - [ ] Even more unit testing, focusing on accurate and human-readable errors.
  
  
##  Not in scope for this particular P.R
(other actions required in other packages.
 - [ ] Issue go-swagger/go-swagger#1308 - Call Log.Fatal() in go-openapi/spec
       More generally, this one is about chasing panic(), Fatal() and so on at the spec level
       (possibly at the loads level)
 - [x]  (closed) Issue go-swagger/go-swagger#14 - duplicate entries in security reference scopes.
       As mentioned with this issue, the problem can't be solved at the validation level.
       If we would like to provide such a nice message, it should be very early in spec parsing.
       I don't think this a workable objective, so OK with having this issue closed.
       DONE: remove stub check entry from code.
 - [ ] Issue go-swagger/go-swagger#1098 - Explicit name collisions: the validate package does already
        a good job at detecting duplicates. This issue is definitely not related to spec validation, but rather
       to a template check that could be performed along the lines of the format check.

### CODE REVIEW

**Comment by fredbi** [comment]
- Experienced a regression in go-openapi/spec today. Possibly lined to go-openapi/spec#47.
- Had to downgrade significantly UT scope.
I gather from build history that I am not the first one (Issue18 test was skipped already).

Running the full suite with -race on my dual-CPU laptop now takes about 1 hour (50 min).
Running it without -race about half this (26 min).

Could we (temporarily) set travis to set a timeout (e.g. go test -timeout 2h) so we may pass at least once the full test suite that I skip again before merging.

Other possibility: removing -race and set -timeout 1h permanently.

I still believe that exercising the validator with a bunch of more or less broken specs (not just low-level unit cases) is healthy.

**Comment by fredbi** [comment]
Observations:
- Test coverage moves into the 90% with the full UT suite
- Patch test coverage should also much better, since most changes have been U-tested with specs and not unitary function calls (except for some edge cases)

**Comment by fredbi** [comment]
You may now exercise go-openapi with extended coverage using go test -v -arg -enable-long

This test option is currently not activated by CI (would require adding -timeout XX or removing -race)

File: rexp.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+// limitations untader the License.
+
+package validate
+
+import (
+	re "regexp"
+)
+
+// Cache of compiled regular expressions
+var reDict = map[string]*re.Regexp{}
```
[the remark this answers is not in the document]
Comment (fredbi) [reply rexp.go]: ok. Got it. Thanks. I'll do that.

Before all, this PR is an investment in UT: I should give us more security for future enhancements.

In particular,validating Examples and stuff will probably need some refactoring. We'll have, I guess, some discussions about it soon.

Also improving the analysis interface as discussed will result in better code here.

[the remark this answers is not in the document]
Comment (fredbi) [reply rexp.go]: Did you run the test -v -coverprofile=cover.out -arg -enable-long ? (without the -race).

It should give you an overview of the actual test coverage which is excellent now.

We miss some fringe cases like default on headers in default response but that's about it.

Also the fail-safe interface with analysis willenable to remove some dead code (checking again and again that $ref are safe...).

[the remark this answers is not in the document]
Comment (fredbi) [reply rexp.go]: Question @casualjim : does the use of sync.Map have an effect on go version compatibility?
We believe we are supposed to support 1.6?

[the remark this answers is not in the document]
Comment (casualjim) [reply rexp.go]: it can also be guarded with a mutex. But we support current go version and 1 version back.

[the remark this answers is not in the document]
Comment (fredbi) [reply rexp.go]: I'd prefer to use a simple sync.Mutex
